### PR TITLE
fix wal not clear and never cleaned[1738]

### PR DESCRIPTION
**Pull request by bartliu827**
# Required checklist
- [ ] Sample config files updated (`config`,`meta/config` and `default config`) 
- [ ] If there are user-facing changes, the documentation needs to be updated prior to approving the PR( [Link]() )
- [ ] If there are any breaking changes to public APIs, please add the `api change` label.
- [ ] Signed [CLA](https://cla-assistant.io/cnosdb/cnosdb) (if not already signed)

# Which issue does this PR close?

https://github.com/cnosdb/cnosdb/issues/1738

[//]: # (We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For -- example `Closes #123` indicates that this PR will close issue #123.)

Related #.

# Rationale for this change

[//]: # (Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.)

# Are there any user-facing changes?

[//]: # (There is no need to duplicate the description in the issue here, but it is sometimes worth providing a summary of the individual changes in this PR.)

### CODE REVIEW

File: tskv/src/wal/mod.rs [hunk 1]
```diff
@@ -473,14 +500,53 @@ impl VnodeWal {
...
+                let wal_files = file_manager::list_file_names(self.wal_dir());
+                for file_name in wal_files {
+                    // If file name cannot be parsed to wal id, skip that file.
+                    let wal_id = match file_utils::get_wal_file_id(&file_name) {
+                        Ok(id) => id,
+                        Err(_) => continue,
+                    };
+
+                    let path = self.wal_dir().join(&file_name);
+                    if let Ok(reader) = WalReader::open(&path).await {
```
Comment (zipper-meng) [inline tskv/src/wal/mod.rs:519]: If it's current writing WAL we don't need to open and check.

File: tskv/src/wal/mod.rs [hunk 2]
```diff
@@ -129,6 +132,7 @@ impl Display for WalType {
             WalType::DeleteTable => write!(f, "delete_table"),
             WalType::UpdateSeriesKeys => write!(f, "update_series_keys"),
             WalType::Delete => write!(f, "delete"),
+            WalType::ClearWalEntry => write!(f, "clear_wal_entry"),
```
Comment (zipper-meng) [inline tskv/src/wal/mod.rs:135]: To **clear wal entry** we didn't write anything to a WAL file, what does WalType::ClearWalEntry do?

File: tskv/src/summary.rs [hunk 3]
```diff
@@ -577,10 +583,25 @@ impl Summary {
...
+
+                if !self.opt.raft_mode {
+                    let (task, rx) = WalTask::new_clear_wal_entry(
+                        tsf.read().await.tenant_database().to_string(),
+                        tsf_id,
+                        last_seq,
+                    );
+
+                    if self.wal_sender.send(task).await.is_ok() {
+                        let _ = rx.await;
```
Comment (zipper-meng) [inline tskv/src/summary.rs:601]: Execution of WalTask::ClearWalEntry opens all WAL files and read the footer from them, I think `rx.await` needs to be out of the loop:
```rs
receivers[]
the loop {
    ( tx, rx ) = get channel
    send wal task ( tx )
    receivers.push(rx)
}
for rx in receivers {
    let_ = rx.await
}
```

File: tskv/src/wal/mod.rs [hunk 4]
```diff
@@ -473,14 +500,53 @@ impl VnodeWal {
...
                     .await
             }
+            writer::Task::ClearWalEntry(ClearWalEntry {
+                tenant_database,
+                sequence,
+                vnode_id,
+                ..
+            }) => {
+                let mut delete_ids = vec![];
+                let wal_files = file_manager::list_file_names(self.wal_dir());
```
Comment (zipper-meng) [inline tskv/src/wal/mod.rs:510]: The map of wal_file_id and it's footer data `(min_seq, max_seq)` could be stored in wal manager when it rolls a wal writer, just as it did in the past versions, then we don't need to open all wal files to get them.

File: tskv/src/wal/mod.rs (hunk 2 above)
Comment (bartliu827) [reply tskv/src/wal/mod.rs:135]: not write,just clear files.

**Comment by zipper-meng** [comment]
The pull request not only resolves WAL issues but also implements API `watch_meta_membership` for singleton mode.

File: tskv/src/summary.rs [hunk 5]
```diff
@@ -577,15 +584,34 @@ impl Summary {
...
+                        tsf.read().await.tenant_database().to_string(),
+                        tsf_id,
+                        last_seq,
+                    );
+
+                    if self.wal_sender.send(task).await.is_ok() {
+                        receivers.push(rx);
+                    }
+                }
+
```
Comment (roseboy-liu) [inline tskv/src/summary.rs:605]: we only need to trigger delete wal file on flush job not on all new version generated.

File: tskv/src/wal/mod.rs (hunk 4 above)
Comment (roseboy-liu) [reply tskv/src/wal/mod.rs:510]: We can optimize next.

**Review by roseboy-liu — APPROVED** [review]

**Pull request merged**